### PR TITLE
Do not load CodeStyle analyzers added by the SDK

### PR DESCRIPTION
**Pull request by JoeRobich**
To avoid a version mismatch between the version of Roslyn in the IDE and the version of Roslyn CodeStyle analyzers added by the SDK (#72811), we will not load the CodeStyle analyzers added by the SDK. The IDE ships with the same analyzers which are version matched and used during analysis.

The analyzers are still listed in the Solution Explorer but without any associated analyzers.
image

### CODE REVIEW

File: src/Workspaces/Core/Portable/Workspace/ProjectSystem/ProjectSystemProject.cs
Comment (ToddGrun) [inline src/Workspaces/Core/Portable/Workspace/ProjectSystem/ProjectSystemProject.cs:1054]: >PathUtilities.DirectorySeparatorStr [](http://example.com/codeflow?start=138&length=35)

Should this be at the beginning of the string too?

File: src/Workspaces/Core/Portable/Workspace/ProjectSystem/ProjectSystemProject.cs
Comment (ToddGrun) [inline src/Workspaces/Core/Portable/Workspace/ProjectSystem/ProjectSystemProject.cs:1060]: >LastIndexOf [](http://example.com/codeflow?start=21&length=11)

EndsWith?

File: src/Workspaces/Core/Portable/Workspace/ProjectSystem/ProjectSystemProject.cs
Comment (ToddGrun) [inline src/Workspaces/Core/Portable/Workspace/ProjectSystem/ProjectSystemProject.cs:1068]: nit: this would be a bit easier to read as a separate fn since it's done in 3 places

File: src/Workspaces/Core/Portable/Workspace/ProjectSystem/ProjectSystemProject.cs
Comment (sharwell) [inline src/Workspaces/Core/Portable/Workspace/ProjectSystem/ProjectSystemProject.cs:1054]: 💭 Should we just look at the assembly name? With the current approach, if users explicitly reference a version of the code style package instead of using the one in the SDK, they will end up in the same bad IDE state we have today.

**Comment by sharwell** [comment]
Note that this will load the analyzers which ship in the IDE, but it does _not_ correct the error where the wrong options are provided to those analyzers (analyzers in the IDE use values from Tools > Options, but those same analyzers in a build do not).

**Review by arkalyanms — APPROVED** [review]

File: src/VisualStudio/Core/Test/ProjectSystemShim/VisualStudioProjectTests/AnalyzerReferenceTests.vb
Comment (jaredpar) [inline src/VisualStudio/Core/Test/ProjectSystemShim/VisualStudioProjectTests/AnalyzerReferenceTests.vb:226]: Are all of these DLLs already in box that we're mapping to?

File: src/Workspaces/Core/Portable/Workspace/ProjectSystem/ProjectSystemProject.cs
Comment (jaredpar) [inline src/Workspaces/Core/Portable/Workspace/ProjectSystem/ProjectSystemProject.cs:1107]: My one fear is that this is hard to test in a unit test. Do we have an item on the queue to do manual verification of Razor / Code style once this is all merged in?

Have we thought about adding an integration test that verifies we redirected these loads?

File: src/VisualStudio/Core/Test/ProjectSystemShim/VisualStudioProjectTests/AnalyzerReferenceTests.vb
Comment (jasonmalinowski) [inline src/VisualStudio/Core/Test/ProjectSystemShim/VisualStudioProjectTests/AnalyzerReferenceTests.vb:218]: Should we just remove this bit since it's not relevant to the test at hand? I imagine this has gotten copy/pastede from prior tests and didn't matter here. It'll also make the assert an Assert.Empty( )which will be a bit clearer.

File: src/VisualStudio/Core/Test/ProjectSystemShim/VisualStudioProjectTests/AnalyzerReferenceTests.vb
Comment (jasonmalinowski) [inline src/VisualStudio/Core/Test/ProjectSystemShim/VisualStudioProjectTests/AnalyzerReferenceTests.vb:226]: Do resource binaries get added too?

Comment (jasonmalinowski) [reply src/VisualStudio/Core/Test/ProjectSystemShim/VisualStudioProjectTests/AnalyzerReferenceTests.vb:226]: Probably needs VB too.

File: src/Workspaces/Core/Portable/Workspace/ProjectSystem/ProjectSystemProject.cs
Comment (jasonmalinowski) [inline src/Workspaces/Core/Portable/Workspace/ProjectSystem/ProjectSystemProject.cs:1059]: Any reason not to just match on Sdks\Microsoft.NET.Sdk\codestyle and work for both languages at once?

File: src/VisualStudio/Core/Test/ProjectSystemShim/VisualStudioProjectTests/AnalyzerReferenceTests.vb
Comment (jasonmalinowski) [inline src/VisualStudio/Core/Test/ProjectSystemShim/VisualStudioProjectTests/AnalyzerReferenceTests.vb:209]:
```suggestion
        Public Async Function CodeStyleAnalyzers_FromSdk_AreIgnored() As Task
```

**Comment by jasonmalinowski** [comment]
> The analyzers are still listed in the Solution Explorer but without any associated analyzers.

Do we have a design-time target somewhere we can add a hack in that removes these at the MSBuild level instead? This is a perfectly fine tactical fix of course, but that'd avoid side effects like this.

File: src/Workspaces/Core/Portable/Workspace/ProjectSystem/ProjectSystemProject.cs
Comment (dibarbet) [inline src/Workspaces/Core/Portable/Workspace/ProjectSystem/ProjectSystemProject.cs:1041]: Will this work as is for VSCode?  I noticed that we don't have a codestyle dll shipped in the extension - do we need that, or do we just need the Features dll?

Comment (jasonmalinowski) [inline src/Workspaces/Core/Portable/Workspace/ProjectSystem/ProjectSystemProject.cs:1041]: Just to call it out explicitly, doing this here will:

1. Also apply to VS Code (which might be reasonable since we'll have similar issues there I imagine)
2. Will _not_ apply to MSBuildWorkspace, which might be bad, or might not be.

The latter we can of course defer for this change (since this has huge dogfooding benefit), but wanted to call out the former.

**Review by jasonmalinowski — APPROVED** [review]

File: src/Workspaces/Core/Portable/Workspace/ProjectSystem/ProjectSystemProject.cs
Comment (JoeRobich) [reply src/Workspaces/Core/Portable/Workspace/ProjectSystem/ProjectSystemProject.cs:1041]: This does work in VS Code. We only need the Features assemblies.

File: src/VisualStudio/Core/Test/ProjectSystemShim/VisualStudioProjectTests/AnalyzerReferenceTests.vb
Comment (JoeRobich) [reply src/VisualStudio/Core/Test/ProjectSystemShim/VisualStudioProjectTests/AnalyzerReferenceTests.vb:226]: I don't believe resource binaries are added as AnalyzerReferences to the project. I would expect they are loaded by the AnalyzerAssemblyLoader.

File: src/Workspaces/Core/Portable/Workspace/ProjectSystem/ProjectSystemProject.cs
Comment (JoeRobich) [reply src/Workspaces/Core/Portable/Workspace/ProjectSystem/ProjectSystemProject.cs:1059]: Happy to make that change. Didn't have a good feel with how specific we wanted to get.

Comment (JoeRobich) [reply src/Workspaces/Core/Portable/Workspace/ProjectSystem/ProjectSystemProject.cs:1059]: Actually will keep it as is. The code currently checks for the analyzer path to end with the folder path fragment which is a stronger guarantee than just containing the fragment.

File: src/VisualStudio/Core/Test/ProjectSystemShim/VisualStudioProjectTests/AnalyzerReferenceTests.vb
Comment (JoeRobich) [reply src/VisualStudio/Core/Test/ProjectSystemShim/VisualStudioProjectTests/AnalyzerReferenceTests.vb]: Removed!

File: src/VisualStudio/Core/Test/ProjectSystemShim/VisualStudioProjectTests/AnalyzerReferenceTests.vb
Comment (JoeRobich) [reply src/VisualStudio/Core/Test/ProjectSystemShim/VisualStudioProjectTests/AnalyzerReferenceTests.vb:218]: These are the in box code style analyzers. We are ensuring that they do not get returned as AnalyzerReferences from the Project. Which means they will not be used for analysis.

File: src/Workspaces/Core/Portable/Workspace/ProjectSystem/ProjectSystemProject.cs
Comment (ToddGrun) [inline src/Workspaces/Core/Portable/Workspace/ProjectSystem/ProjectSystemProject.cs:1097]: >" " [](http://example.com/codeflow?start=109&length=3)

Why did this switch from the directory separator char to a space?

Comment (JoeRobich) [reply src/Workspaces/Core/Portable/Workspace/ProjectSystem/ProjectSystemProject.cs:1097]: So Path.Combine will add the separator then we trim it off the spaces. Was just trying to put the work onto Path.Combine.

Comment (JoeRobich) [reply src/Workspaces/Core/Portable/Workspace/ProjectSystem/ProjectSystemProject.cs:1097]: Maybe it is dumb. Happy to revert it.

Comment (ToddGrun) [reply src/Workspaces/Core/Portable/Workspace/ProjectSystem/ProjectSystemProject.cs:1097]: Definitely not dumb, just a little too tricky for me to spot without the explanation.

File: src/Workspaces/Core/Portable/Workspace/ProjectSystem/ProjectSystemProject.cs
Comment (JoeRobich) [reply src/Workspaces/Core/Portable/Workspace/ProjectSystem/ProjectSystemProject.cs:1091]: I think it should be easy enough to write an integration test for this. Will do as a follow up after the snap.

File: src/Workspaces/Core/Portable/Workspace/ProjectSystem/ProjectSystemProject.cs
Comment (JoeRobich) [reply src/Workspaces/Core/Portable/Workspace/ProjectSystem/ProjectSystemProject.cs:1097]: I am going to do a follow up that adds an integration test. Will add a comment here as well.

File: src/Workspaces/Core/Portable/Workspace/ProjectSystem/ProjectSystemProject.cs
Comment (chsienki) [reply src/Workspaces/Core/Portable/Workspace/ProjectSystem/ProjectSystemProject.cs:1041]: I really don't love that we're expanding this mechanism, given the work we're doing to allow the SDK to redirect analyzers via https://github.com/dotnet/roslyn/pull/74820 

That isn't a reason to block on this, given its not done yet, but could we open an issue to investigate if we could replace this with the new API when it's introduced?

**Comment by KirillOsenkov** [comment]
@JoeRobich quick question: will it work if the paths to analyzers contain `..`?

I haven't tried in VS, but some other hosts receive paths like this:
`C:\Program Files\dotnet\sdk\9.0.100-rc.1.24452.12\Sdks\Microsoft.NET.Sdk\targets\..\codestyle\cs\Microsoft.CodeAnalysis.CodeStyle.dll`, so need to first call `Path.GetFullPath()` on the reference before calling IsSdk...

I haven't checked whether the VS project system normalizes paths before this codepath is hit.

**Comment by JoeRobich** [comment]
> so need to first call `Path.GetFullPath()` on the reference before calling IsSdk...

Thanks for checking! We do call `Path.GetFullPath()` just before.

https://github.com/dotnet/roslyn/blob/3bff3622487486dec7794dfd0c71e05a52c313a4/src/Workspaces/Core/Portable/Workspace/ProjectSystem/ProjectSystemProject.cs#L1033-L1037

**Pull request merged**
